runtime/compiler/optimizer/InterpreterEmulator.cpp

```diff
@@ -0,0 +1,856 @@
+#include "optimizer/InterpreterEmulator.hpp"
+#include "optimizer/J9EstimateCodeSize.hpp"
+#include "env/VMAccessCriticalSection.hpp"
+#include "optimizer/PreExistence.hpp"
+#include "optimizer/J9CallGraph.hpp"
+#include "ilgen/IlGenRequest.hpp"
+#include "jilconsts.h"
+
+void
+InterpreterEmulator::maintainStackForIf(TR_J9ByteCode bc)
+   {
+   TR_ASSERT_FATAL(_iteratorWithState, "has to be called when the iterator has state!");
+   int32_t branchBC = _bcIndex + next2BytesSigned();
+   int32_t fallThruBC = _bcIndex + 3;
+   IconstOperand * second = pop()->asIconst();
+   IconstOperand * first = pop()->asIconst();
+   bool canBranch = true;
+   bool canFallThru = true;
+   if (second && first)
+      {
+      switch (bc)
+         {
+         case J9BCificmpeq:
+            canBranch = second->intValue == first->intValue;
+            debugTrace(tracer(), "maintainStackForIf ifcmpeq %d == %d\n", second->intValue, first->intValue);
+            break;
+         case J9BCificmpne:
+            canBranch = second->intValue != first->intValue;
+            debugTrace(tracer(), "maintainStackForIf ifcmpne %d != %d\n", second->intValue, first->intValue);
+            break;
+         }
+      canFallThru = !canBranch;
+      }
+
+   if (canBranch)
+      {
+      debugTrace(tracer(), "maintainStackForIf canBranch to bcIndex=%d\n", branchBC);
+      genTarget(branchBC);
+      }
+
+   if (canFallThru)
+      {
+      debugTrace(tracer(), "maintainStackForIf canFallThrough to bcIndex=%d\n", fallThruBC);
+      genTarget(fallThruBC);
+      }
+   }
+
+void
+InterpreterEmulator::maintainStackForGetField()
+   {
+   TR_ASSERT_FATAL(_iteratorWithState, "has to be called when the iterator has state!");
+   bool isVolatile, isPrivate, isUnresolvedInCP, isFinal;
+   TR::DataType type = TR::NoType;
+   uint32_t fieldOffset;
+   int32_t cpIndex = next2Bytes();
+   Operand *newOperand = _unknownOperand;
+   bool resolved = _calltarget->_calleeMethod->fieldAttributes(comp(), cpIndex, &fieldOffset, &type, &isVolatile, &isFinal, &isPrivate, false, &isUnresolvedInCP, false);
+   if (top()->getKnownObjectIndex() != TR::KnownObjectTable::UNKNOWN && type == TR::Address)
+      {
+      TR::Symbol::RecognizedField recognizedField = TR::Symbol::searchRecognizedField(comp(), _calltarget->_calleeMethod, cpIndex, false);
+      TR::Symbol *fieldSymbol = NULL;
+      if (recognizedField != TR::Symbol::UnknownField)
+         fieldSymbol = TR::Symbol::createRecognizedShadow(trStackMemory(),type, recognizedField);
+      else
+         fieldSymbol = TR::Symbol::createShadow(trStackMemory(),type);
+      if (isFinal)
+         fieldSymbol->setFinal();
+
+      if ((resolved || !isUnresolvedInCP) && comp()->fej9()->canDereferenceAtCompileTimeWithFieldSymbol(fieldSymbol, cpIndex, _calltarget->_calleeMethod))
+         {
+         TR::VMAccessCriticalSection dereferenceKnownObjectField(comp()->fej9());
```

Comment:
Shouldn't we check for the knot before we acquire the lock to save overhead if the knot is disabled?

Reply:
changed the code to check the `knot` first